### PR TITLE
T13672 factories

### PR DESCRIPTION
**Pull request by niden**
Hello!

* Type: bug fix/new feature
* Link to issue: #13672 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [x] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

## Added
- Added `Phalcon\Factory\Exception` for factory exceptions.
- Added `Phalcon\Filter\FilterInterface` for custom filter implementations. 
- Added `Phalcon\Db\Adapter\PdoFactory`: Factory to create PDO adapters. 
- Added `Phalcon\Image\ImageFactory`: Factory to create image adapters.
- Added `Phalcon\Image\LoggerFactory`: Factory to create logger objects.
- Added `Phalcon\Translate\InterpolatorFactory`: Factory to create interpolator objects. 
- Added `Phalcon\Translate\TranslateFactory`: Factory to create translate objects. 

## Changed
- Renamed `Phalcon\Annotations\Adapter\Files` to `Phalcon\Annotations\Adapter\Stream`. 
- Renamed `Phalcon\Annotations\Adapter` to `Phalcon\Annotations\Adapter\AbstractAdapter`. 
- Renamed `Phalcon\Annotations\AdapterInterface` to `Phalcon\Annotations\Adapter\AdapterInterface`. [#13672](https://github.com/phalcon/cphalcon/issues/13672)
- Renamed `Phalcon\Annotations\Factory` to `Phalcon\Annotations\AnnotationsFactory`.
- Renamed `Phalcon\Config\Factory` to `Phalcon\Config\ConfigFactory`. 
- Renamed `Phalcon\Filter\FilterLocator` to `Phalcon\Filter\Filter`. 
- Renamed `Phalcon\Filter\FilterLocatorFactory` to `Phalcon\Filter\FilterFactory`. 
- Renamed `Phalcon\Image\Adapter` to `Phalcon\Image\Adapter\AbstractAdapter`. 
- Renamed `Phalcon\Image\AdapterInterface` to `Phalcon\Image\Adapter\AdapterInterface`.
- Renamed `Phalcon\Paginator\Adapter` to `Phalcon\Paginator\Adapter\AbstractAdapter`. 
- Renamed `Phalcon\Paginator\AdapterInterface` to `Phalcon\Paginator\Adapter\AdapterInterface`. 
- Renamed `Phalcon\Paginator\Factory` to `Phalcon\Paginator\PaginatorFactory`. 

## Removed
- Removed `Phalcon\Session\Factory`. 
- Removed `Phalcon\Factory` and `Phalcon\FactoryInterface`. 

Thanks

### CODE REVIEW

File: phalcon/Annotations/Adapter/Stream.zep [hunk 1]
```diff
@@ -20,7 +20,7 @@ use Phalcon\Annotations\Exception;
  * Stores the parsed annotations in files. This adapter is suitable for production
  *
  *<code>
- * use Phalcon\Annotations\Adapter\Files;
+ * use Phalcon\Annotations\Adapter\Stream;
  *
  * $annotations = new Files(
```
Comment (sergeyklay) [inline phalcon/Annotations/Adapter/Stream.zep]:
```suggestion
 * $annotations = new Stream(
```

Comment (sergeyklay) [reply phalcon/Annotations/Adapter/Stream.zep]: - At the top of class annotation: `Phalcon\Annotations\Adapter\Files` => `Phalcon\Annotations\Adapter\Stream`
- Constructor annotation: `Phalcon\Annotations\Adapter\Files constructor` => `Phalcon\Annotations\Adapter\Stream constructor`

File: phalcon/Config/ConfigFactory.zep [hunk 2]
```diff
@@ -80,7 +74,7 @@ class Factory extends BaseFactory
             );
         }
 
-        let className = namespaceClass . "\\" . camelize(adapter);
+        let className = "Phalcon\\Config\\Adapter\\" . camelize(adapter);
```
Comment (sergeyklay) [inline phalcon/Config/ConfigFactory.zep]: What about user defined classes?

File: phalcon/Factory/AbstractFactory.zep [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Returns the adapters for the factory
+     */
+    abstract protected function getAdapters() -> array;
+
+    /**
+     * AdapterFactory constructor.
+     */
+    protected function init(array! services = [])
```
Comment (sergeyklay) [inline phalcon/Factory/AbstractFactory.zep]:
```suggestion
    protected function init(array! services = []) -> void
```

**Review by sergeyklay — CHANGES_REQUESTED** [review]

File: phalcon/Annotations/Adapter/Stream.zep (hunk 1 above)
Comment (niden) [reply phalcon/Annotations/Adapter/Stream.zep]: Fixed

File: phalcon/Config/ConfigFactory.zep (hunk 2 above)
Comment (niden) [reply phalcon/Config/ConfigFactory.zep]: This was never designed for custom classes for config. The old `load` method would load only a prefix namespace and then one of the adapters

File: phalcon/Factory/AbstractFactory.zep (hunk 3 above)
Comment (niden) [reply phalcon/Factory/AbstractFactory.zep]: Fixed

**Comment by niden** [comment]
@sergeyklay see my response above. Also I will regenerate `ext/` after this is merged

**Review by sergeyklay — APPROVED** [review]

**Pull request merged**